test/close.test.js

```diff
@@ -4,290 +4,355 @@ const net = require('net')
+//         t.match(data.toString(), /Connection:\s*keep-alive/i)
+//         t.match(data.toString(), /200 OK/i)
+
+//         client.write('GET / HTTP/1.1\r\n\r\n')
+
+//         client.once('data', data => {
+//           t.match(data.toString(), /Connection:\s*close/i)
+//           t.match(data.toString(), /200 OK/i)
+
+//           // Test that fastify closes the TCP connection
+//           client.once('close', () => {
+//             t.end()
+//           })
+//         })
+//       })
+//     })
+//   })
+// })
+
+// t.test('Current opened connection should not accept new incoming connections', t => {
+//   t.plan(3)
+//   const fastify = Fastify()
+//   fastify.get('/', (req, reply) => {
+//     fastify.close()
+//     setTimeout(() => {
+//       reply.send({ hello: 'world' })
+//     }, 250)
+//   })
+
+//   fastify.listen(0, err => {
+//     t.error(err)
+//     const instance = new Client('http://localhost:' + fastify.server.address().port)
+//     instance.request({ path: '/', method: 'GET' }).then(data => {
+//       t.equal(data.statusCode, 200)
+//     })
+//     instance.request({ path: '/', method: 'GET' }).then(data => {
+//       t.equal(data.statusCode, 503)
+//     })
+//   })
+// })
+
+// test('Cannot be reopened the closed server without listen callback', async t => {
+//   t.plan(2)
+//   const fastify = Fastify()
+
+//   await fastify.listen(0)
+//   await fastify.close()
+
+//   try {
+//     await fastify.listen(0)
+//   } catch (err) {
+//     t.ok(err)
+//     t.equal(err.code, 'FST_ERR_REOPENED_CLOSE_SERVER')
+//   }
+// })
+
+// test('Cannot be reopened the closed server has listen callback', async t => {
+//   t.plan(2)
+//   const fastify = Fastify()
+
+//   await fastify.listen(0)
+//   await fastify.close()
+
+//   await new Promise((resolve, reject) => {
+//     fastify.listen(0, err => {
+//       reject(err)
+//     })
+//   }).catch(err => {
+//     t.equal(err.code, 'FST_ERR_REOPENED_CLOSE_SERVER')
+//     t.ok(err)
+//   })
+// })
+
+// test('shutsdown while keep-alive connections are active', async t => {
+//   t.plan(5)
+
+//   const timeoutTime = 2 * 60 * 1000
+//   const fastify = Fastify({ forceCloseConnections: true })
+
+//   fastify.server.setTimeout(timeoutTime)
+//   fastify.server.keepAliveTimeout = timeoutTime
+
+//   fastify.get('/', (req, reply) => {
+//     reply.send({ hello: 'world' })
+//   })
+
+//   const address = await fastify.listen(0)
+//   const client = new Client(address, { keepAliveTimeout: 1 * 60 * 1000 })
+
+//   const response = await client.request({ path: '/', method: 'GET' })
+//   const body = await response.body.text()
+//   t.equal(body, '{"hello":"world"}')
+//   t.equal(client.closed, false)
+
+//   client.on('disconnect', () => {
+//     t.pass()
+//   })
+
+//   fastify.close((err) => {
+//     t.error(err)
+//     t.equal(client.closed, true)
+//   })
+// })
+
+test('shutsdown while keep-alive connections are active (non-async)', t => {
```

Comment:
I am very confused by this test. Set a breakpoint in the `fastify.close` callback and it will never be hit. At the same time, set a breakpoint on line 384 of `fastify.js` (the loop that unrefs the keep-alive sockets) and notice that it gets hit _twice_.

Comment out this test, and uncomment any other original test, and notice that the loop breakpoint only gets hit once.

There is something wrong in this test, and I'm at a loss for what it is.

Reply:
🤦‍♂️ the loop being hit twice is just the natural iteration of a loop with a single item in the iterable. Still, no idea why this thing is not invoking the callback passed to `fastify.close`.